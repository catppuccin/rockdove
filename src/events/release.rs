use octocrab::models::webhook_events::{
    payload::{ReleaseWebhookEventAction, ReleaseWebhookEventPayload},
    WebhookEvent,
};

use crate::{colors::RELEASE_COLOR, embed_builder::EmbedBuilder};

pub fn make_release_embed(
    event: WebhookEvent,
    specifics: &ReleaseWebhookEventPayload,
) -> Option<EmbedBuilder> {
    if !matches!(specifics.action, ReleaseWebhookEventAction::Released) {
        return None;
    }

    let repo = event
        .repository
        .expect("release events should always have a repository");

    let mut embed = EmbedBuilder::default();

    let repo_name = repo.full_name.unwrap_or(repo.name);

    embed.title(&format!(
        "[{}] New release published: {}",
        repo_name,
        specifics
            .release
            .get("name")
            .and_then(|v| v.as_str())
            .unwrap_or("*no name*"),
    ));

    embed.url(
        specifics
            .release
            .get("html_url")
            .and_then(|v| v.as_str())
            .expect("release should always have an html url"),
    );

    if let Some(body) = specifics.release.get("body").and_then(|v| v.as_str()) {
        embed.description(body);
    }

    embed.color(RELEASE_COLOR);

    Some(embed)
}

#[cfg(test)]
mod tests {
    use crate::{
        make_embed,
        tests::{embed_context, TestConfig},
    };

    #[test]
    fn released() {
        let payload = include_str!("../../fixtures/release/released.json");
        let TestConfig {
            webhook_event: event,
            mut settings,
        } = TestConfig::new("release", payload);

        let embed = make_embed(event)
            .expect("make_embed should succeed")
            .expect("event fixture can be turned into an embed");

        settings.set_info(&embed_context(&embed));
        settings.bind(|| insta::assert_yaml_snapshot!(embed));
    }
}
