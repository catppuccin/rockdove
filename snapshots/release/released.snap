---
source: src/events/release.rs
assertion_line: 71
info:
  author_name_length: 8
  colour_hex: "#CBA6F7"
  description_length: 12
  title_length: 54
---
embeds:
  - author:
      icon_url: "https://avatars.githubusercontent.com/u/58985301?v=4"
      name: sgoudham
      url: "https://github.com/sgoudham"
    color: 13346551
    description: Test release
    title: "[catppuccin-rfc/polybar] New release published: v0.0.0"
    url: "https://github.com/catppuccin-rfc/polybar/releases/tag/v0.0.0"
